framework/src/outputs/RegurgitateInput.C

```diff
@@ -0,0 +1,204 @@
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+// MOOSE includes
+#include "RegurgitateInput.h"
+#include "InputParameterWarehouse.h"
+#include "FEProblemBase.h"
+#include "Conversion.h"
+#include "parse.h"
+
+#include "SetAdaptivityOptionsAction.h"
+#include "CreateProblemAction.h"
+#include "AddVariableAction.h"
+#include "AddAuxVariableAction.h"
+
+template <>
+InputParameters
+validParams<RegurgitateInput>()
+{
+  InputParameters params = validParams<FileOutput>();
+  params.set<ExecFlagEnum>("execute_on") = EXEC_INITIAL;
+  params.suppressParameter<ExecFlagEnum>("execute_on");
+  params.addParam<bool>("regurgitate", false, "Disable the input file reguritatation.");
+  params.addParam<bool>(
+      "show_on_console", false, "Enable the display of the reguritated input on the screen.");
+  return params;
+}
+
+RegurgitateInput::RegurgitateInput(const InputParameters & parameters) : FileOutput(parameters)
+{
+  // Create a map from "_moose_base" to the necessary syntax for re-creating input file. The
+  // pair contains two pieces of information: the parent input file syntax for the parameters
+  // and a bool indicating if the object is created in a sub-block (true).
+  // TODO: There should be a way to get this info from the action warehouse
+  _base_to_syntax["Output"] = std::make_pair("Outputs", true);
+  _base_to_syntax["Kernel"] = std::make_pair("Kernels", true);
+  _base_to_syntax["AuxKernel"] = std::make_pair("AuxKernels", true);
+  _base_to_syntax["ScalarKernel"] = std::make_pair("ScalarKernels", true);
+  _base_to_syntax["AuxScalarKernel"] = std::make_pair("AuxScalarKernels", true);
+  _base_to_syntax["Marker"] = std::make_pair("Adaptivity/Markers", true);
+  _base_to_syntax["BoundaryCondition"] = std::make_pair("BCs", true);
+  _base_to_syntax["Function"] = std::make_pair("Functions", true);
+  _base_to_syntax["InitialCondition"] = std::make_pair("ICs", true);
+  _base_to_syntax["ScalarInitialCondition"] = std::make_pair("ScarlarICs", true);
+  _base_to_syntax["Postprocessor"] = std::make_pair("Postprocessors", true);
+  _base_to_syntax["VectorPostprocessor"] = std::make_pair("VectorPostprocessors", true);
+  _base_to_syntax["MoosePreconditioner"] = std::make_pair("Preconditioner", true);
+  _base_to_syntax["Material"] = std::make_pair("Materials", true);
+
+  _base_to_syntax["Executioner"] = std::make_pair("Executioner", false);
+  _base_to_syntax["TimeStepper"] = std::make_pair("Executioner/TimeStepper", false);
+  _base_to_syntax["TimeIntegrator"] = std::make_pair("Executioner/TimeIntegrator", false);
+
+  _base_to_syntax["MooseMesh"] = std::make_pair("Mesh", false);
+  _base_to_syntax["Problem"] = std::make_pair("Problem", false);
```

Comment:
So I gave this more thought and realized that what you want out of the syntax object is kind of there, but kind of not. First off, we really don't want to put Syntax in this object because it'll definitely get out of sync with the syntax object. So the reason it's difficult to get the association from an object to it's syntax is because there's a one to many relationship. A single piece of syntax can be associated with many Actions which can be setup to create many different types of objects.
E.g.
`Constraints` are created from Actions associated both the `Constraints/*` block as well as the `ContactAction` block. Users are also free to create more blocks where Constraints can be created as well.

So even if we could get back this information directly, it would be a list of possible blocks that can create these object types.

Solution:
I believe the best way to handle this would be to add another optional parameter to the Syntax registration that takes a Boolean indicating whether the block is the "primary" or "preferred" syntax block. Then when a developer registers a new syntax, they'll be able to set whether or not the registered block is the primary blocks and we can use that information to return a single piece of syntax for a given object.

Reply:
How about along with "_moose_base" we add another "_moose_syntax" or something that tells us what syntax will build the object.

This will almost always be at the MOOSE level so it shouldn't be that hard to maintain, if you add a new system then you need "_moose_base" and "_moose_syntax".

Reply:
Nope! That means we'll be bringing in and hardcoding the syntax right inside every pluggable system. If somebody wants to dump our developer syntax, they'd be screwed. If it's in the syntax object, it's subject to being overridable by applications. I think just another method or argument to the existing calls in Syntax will do what we need.